ext/opcache/Optimizer/sccp.c

```diff
@@ -0,0 +1,1463 @@
+		 * CV-affecting instructions into CONST ASSIGNs. This basic DCE is performed for multiple reasons:
+		 * a) During operand replacement we eliminate FREEs. The corresponding computational instructions
+		 *    must be removed to avoid leaks. This way SCCP can run independently of the full DCE pass.
+		 * b) The main DCE pass relies on type analysis to determine whether instructions have side-effects
+		 *    and can't be DCEd. This means that it will not be able collect all instructions rendered dead
+		 *    by SCCP, because they may have potentially side-effecting types, but the actual values are
+		 *    not. As such doing DCE here will allow us to eliminate more dead code in combination.
+		 * c) The ordinary DCE pass cannot collect dead calls. However SCCP can result in dead calls, which
+		 *    we need to collect. */
+
+		if (var->definition >= 0 && value_known(&ctx->values[i])) {
+			zend_op *opline = &op_array->opcodes[var->definition];
+			zend_ssa_op *ssa_op = &ssa->ops[var->definition];
+			if (opline->opcode == ZEND_ASSIGN) {
+				/* Leave assigns to DCE (due to dtor effects) */
+				continue;
+			}
+
+			if (ssa_op->result_def == i
+					&& ssa_op->op1_def < 0
+					&& ssa_op->op2_def < 0
+					&& var->use_chain < 0
+					&& var->phi_use_chain == NULL) {
+				if (opline->opcode == ZEND_DO_ICALL) {
+					/* Call instruction -> remove opcodes that are part of the call */
+					zend_call_info *call;
+					int i;
+
+					ZEND_ASSERT(ctx->call_map);
+					call = ctx->call_map[var->definition];
+					ZEND_ASSERT(call);
+					ZEND_ASSERT(call->caller_call_opline == opline);
+					if (opline->result_type & (IS_TMP_VAR|IS_VAR)) {
+						zend_optimizer_remove_live_range_ex(op_array, opline->result.var, var->definition + 1);
+					}
+					zend_ssa_remove_result_def(ssa, ssa_op);
+					zend_ssa_remove_instr(ssa, opline, ssa_op);
+					zend_ssa_remove_instr(ssa, call->caller_init_opline,
+						&ssa->ops[call->caller_init_opline - op_array->opcodes]);
+
+					for (i = 0; i < call->num_args; i++) {
+						zend_ssa_remove_instr(ssa, call->arg_info[i].opline,
+							&ssa->ops[call->arg_info[i].opline - op_array->opcodes]);
+					}
+					removed_ops = call->num_args + 2;
+
+					// TODO: remove call_info completely???
+					call->callee_func = NULL;
+				} else {
+					/* Ordinary computational instruction -> remove it */
+					if (opline->result_type & (IS_TMP_VAR|IS_VAR)) {
+						zend_optimizer_remove_live_range_ex(op_array, opline->result.var, var->definition + 1);
+					}
+					zend_ssa_remove_result_def(ssa, ssa_op);
+					zend_ssa_remove_instr(ssa, opline, ssa_op);
+					removed_ops++;
+				}
+			} else if (ssa_op->op1_def == i) {
+				/* Compound assign or incdec -> convert to direct ASSIGN */
+
+				/* Destroy previous op2 */
+				if (opline->op2_type == IS_CONST) {
+					literal_dtor(&ZEND_OP2_LITERAL(opline));
+				} else if (ssa_op->op2_use >= 0) {
+					zend_ssa_unlink_use_chain(ssa, var->definition, ssa_op->op2_use);
```

Comment:
This needs a check for `ssa_op->op1_use != ssa_op->op2_use`. Otherwise we fail on code like `$a = 1; $a += $a;`.